File: tests/test_tool_adapter.py
"""Tests for dispatch/tool_adapter.py — schema, CRUD, validation, stub execution, thread safety."""

import sys
import threading
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.tool_adapter import (
    TOOL_ADAPTER_SCHEMA_VERSION,
    ToolDefinition,
    ToolExecutionRequest,
    ToolExecutionResult,
    ToolAdapterManager,
    make_tool,
)


class SchemaVersionTests(unittest.TestCase):
    def test_schema_version(self):
        self.assertEqual(TOOL_ADAPTER_SCHEMA_VERSION, "tool_adapter.v1")

    def test_default_schema_version(self):
        t = make_tool()
        self.assertEqual(t.schema_version, TOOL_ADAPTER_SCHEMA_VERSION)


class DataclassTests(unittest.TestCase):
    def test_tool_definition_fields(self):
        t = make_tool(
            tool_id="web-search", name="Web Search",
            description="Search the web",
            input_schema={"type": "object", "properties": {"query": {"type": "string"}}},
            output_schema={"type": "object", "properties": {"results": {"type": "array"}}},
            timeout_seconds=15,
            requires_network=True,
        )
        self.assertEqual(t.tool_id, "web-search")
        self.assertEqual(t.name, "Web Search")
        self.assertEqual(t.description, "Search the web")
        self.assertTrue(t.requires_network)
        self.assertEqual(t.timeout_seconds, 15)

    def test_tool_definition_defaults(self):
        t = make_tool()
        self.assertEqual(t.timeout_seconds, 30)
        self.assertFalse(t.requires_network)

    def test_frozen(self):
        t = make_tool()
        with self.assertRaises(AttributeError):
            t.name = "changed"

    def test_execution_request(self):
        req = ToolExecutionRequest(
            tool_id="t1", arguments={"q": "test"}, request_id="r1",
        )
        self.assertEqual(req.tool_id, "t1")
        self.assertEqual(req.arguments, {"q": "test"})
        self.assertEqual(req.request_id, "r1")

    def test_execution_result(self):
        res = ToolExecutionResult(
            request_id="r1", tool_id="t1", success=True,
            output={"count": 5}, error=None, duration_ms=1.2,
        )
        self.assertTrue(res.success)
        self.assertEqual(res.output, {"count": 5})


class RegisterUnregisterTests(unittest.TestCase):
    def test_register_valid(self):
        mgr = ToolAdapterManager()
        self.assertTrue(mgr.register_tool(make_tool()))
        self.assertEqual(len(mgr.list_tools()), 1)

    def test_register_duplicate_rejected(self):
        mgr = ToolAdapterManager()
        self.assertTrue(mgr.register_tool(make_tool()))
        self.assertFalse(mgr.register_tool(make_tool()))
        self.assertEqual(len(mgr.list_tools()), 1)

    def test_register_invalid_rejected(self):
        mgr = ToolAdapterManager()
        self.assertFalse(mgr.register_tool(make_tool(tool_id="")))
        self.assertEqual(len(mgr.list_tools()), 0)

    def test_unregister_existing(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool())
        self.assertTrue(mgr.unregister_tool("test-tool"))
        self.assertEqual(len(mgr.list_tools()), 0)

    def test_unregister_nonexistent(self):
        mgr = ToolAdapterManager()
        self.assertFalse(mgr.unregister_tool("nope"))

    def test_get_tool(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool())
        self.assertIsNotNone(mgr.get_tool("test-tool"))
        self.assertIsNone(mgr.get_tool("nope"))

    def test_list_tools(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool(tool_id="a", name="A"))
        mgr.register_tool(make_tool(tool_id="b", name="B"))
        tools = mgr.list_tools()
        self.assertEqual(len(tools), 2)
        ids = {t.tool_id for t in tools}
        self.assertEqual(ids, {"a", "b"})

    def test_list_returns_copy(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool())
        tools = mgr.list_tools()
        tools.clear()
        self.assertEqual(len(mgr.list_tools()), 1)


class ValidationTests(unittest.TestCase):
    def test_valid_tool_no_errors(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool())
        self.assertEqual(errors, [])

    def test_missing_tool_id(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(tool_id=""))
        self.assertTrue(any("tool_id" in e for e in errors))

    def test_missing_name(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(name=""))
        self.assertTrue(any("name" in e for e in errors))

    def test_missing_description(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(description=""))
        self.assertTrue(any("description" in e for e in errors))

    def test_zero_timeout(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(timeout_seconds=0))
        self.assertTrue(any("timeout" in e for e in errors))

    def test_negative_timeout(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(timeout_seconds=-5))
        self.assertTrue(any("timeout" in e for e in errors))

    def test_invalid_schema_version(self):
        mgr = ToolAdapterManager()
        errors = mgr.validate_tool(make_tool(schema_version="wrong.v1"))
        self.assertTrue(any("schema_version" in e for e in errors))


class StubExecutionTests(unittest.TestCase):
    def test_execute_registered_tool(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool())
        req = ToolExecutionRequest(tool_id="test-tool", arguments={"q": "hi"}, request_id="r1")
        result = mgr.execute_tool(req)
        self.assertTrue(result.success)
        self.assertEqual(result.output, {})
        self.assertIsNone(result.error)
        self.assertEqual(result.request_id, "r1")
        self.assertEqual(result.tool_id, "test-tool")
        self.assertGreaterEqual(result.duration_ms, 0)

    def test_execute_unknown_tool(self):
        mgr = ToolAdapterManager()
        req = ToolExecutionRequest(tool_id="missing", arguments={}, request_id="r2")
        result = mgr.execute_tool(req)
        self.assertFalse(result.success)
        self.assertIsNone(result.output)
        self.assertIn("not found", result.error)
        self.assertEqual(result.request_id, "r2")

    def test_execute_after_unregister(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool())
        mgr.unregister_tool("test-tool")
        req = ToolExecutionRequest(tool_id="test-tool", arguments={}, request_id="r3")
        result = mgr.execute_tool(req)
        self.assertFalse(result.success)
        self.assertIn("not found", result.error)

    def test_execute_multiple_tools(self):
        mgr = ToolAdapterManager()
        mgr.register_tool(make_tool(tool_id="t1", name="Tool 1"))
        mgr.register_tool(make_tool(tool_id="t2", name="Tool 2"))
        r1 = mgr.execute_tool(ToolExecutionRequest(tool_id="t1", arguments={}, request_id="r1"))
        r2 = mgr.execute_tool(ToolExecutionRequest(tool_id="t2", arguments={}, request_id="r2"))
        self.assertTrue(r1.success)
        self.assertTrue(r2.success)
        self.assertEqual(r1.tool_id, "t1")
        self.assertEqual(r2.tool_id, "t2")


class ThreadSafetyTests(unittest.TestCase):
    def test_concurrent_register_and_execute(self):
        mgr = ToolAdapterManager()
        errors = []

        def register_many(start: int, count: int) -> None:
            try:
                for i in range(start, start + count):
                    tool = make_tool(tool_id=f"t-{i}", name=f"Tool {i}")
                    mgr.register_tool(tool)
            except Exception as e:
                errors.append(str(e))

        def execute_loop() -> None:
            try:
                for i in range(50):
                    req = ToolExecutionRequest(
                        tool_id=f"t-{i % 100}", arguments={}, request_id=f"r-{i}",
                    )
                    mgr.execute_tool(req)
            except Exception as e:
                errors.append(str(e))

        def list_loop() -> None:
            try:
                for _ in range(50):
                    mgr.list_tools()
            except Exception as e:
                errors.append(str(e))

        threads = [
            threading.Thread(target=register_many, args=(0, 50)),
            threading.Thread(target=register_many, args=(50, 50)),
            threading.Thread(target=execute_loop),
            threading.Thread(target=list_loop),
        ]
        for t in threads:
            t.start()
        for t in threads:
            t.join()

        self.assertEqual(errors, [])
        self.assertEqual(len(mgr.list_tools()), 100)


if __name__ == "__main__":
    unittest.main()

File: engine/tests/test_tool_adapter.rs
use engine::ecosystem::tool_adapter::*;
use std::collections::HashMap;

fn make_test_tool() -> ToolDefinition {
    ToolDefinition {
        schema_version: TOOL_ADAPTER_SCHEMA_VERSION.to_string(),
        tool_id: "t1".to_string(),
        name: "Calculator".to_string(),
        description: "A simple calculator".to_string(),
        input_schema: serde_json::json!({"type": "object"}),
        output_schema: serde_json::json!({"type": "object"}),
        timeout_seconds: 30,
        requires_network: false,
    }
}

#[test]
fn test_register_and_list() {
    let mut mgr = ToolAdapterManager::new();
    assert!(mgr.register_tool(&make_test_tool()));
    assert_eq!(mgr.list_tools().len(), 1);
}

#[test]
fn test_register_duplicate() {
    let mut mgr = ToolAdapterManager::new();
    assert!(mgr.register_tool(&make_test_tool()));
    assert!(!mgr.register_tool(&make_test_tool()));
}

#[test]
fn test_unregister() {
    let mut mgr = ToolAdapterManager::new();
    mgr.register_tool(&make_test_tool());
    assert!(mgr.unregister_tool("t1"));
    assert!(mgr.get_tool("t1").is_none());
    assert!(!mgr.unregister_tool("nonexistent"));
}

#[test]
fn test_validate_valid() {
    let mgr = ToolAdapterManager::new();
    let errors = mgr.validate_tool(&make_test_tool());
    assert!(errors.is_empty());
}

#[test]
fn test_validate_missing_id() {
    let mgr = ToolAdapterManager::new();
    let mut tool = make_test_tool();
    tool.tool_id = String::new();
    let errors = mgr.validate_tool(&tool);
    assert!(errors.iter().any(|e| e.contains("tool_id")));
}

#[test]
fn test_validate_negative_timeout() {
    let mgr = ToolAdapterManager::new();
    let mut tool = make_test_tool();
    tool.timeout_seconds = -1;
    let errors = mgr.validate_tool(&tool);
    assert!(errors.iter().any(|e| e.contains("timeout")));
}

#[test]
fn test_execute_existing_tool() {
    let mut mgr = ToolAdapterManager::new();
    mgr.register_tool(&make_test_tool());
    let request = ToolExecutionRequest {
        tool_id: "t1".to_string(),
        arguments: serde_json::json!({}),
        request_id: "req-1".to_string(),
    };
    let result = mgr.execute_tool(&request, 1000.0);
    assert!(result.success);
    assert!(result.error.is_none());
    assert_eq!(result.tool_id, "t1");
}

#[test]
fn test_execute_missing_tool() {
    let mgr = ToolAdapterManager::new();
    let request = ToolExecutionRequest {
        tool_id: "nonexistent".to_string(),
        arguments: serde_json::json!({}),
        request_id: "req-1".to_string(),
    };
    let result = mgr.execute_tool(&request, 1000.0);
    assert!(!result.success);
    assert!(result.error.is_some());
}

#[test]
fn test_to_dict() {
    let tool = make_test_tool();
    let d = tool.to_dict();
    assert_eq!(d["tool_id"], "t1");
    assert_eq!(d["name"], "Calculator");
}

#[test]
fn test_make_tool_defaults() {
    let tool = make_tool(HashMap::new());
    assert_eq!(tool.tool_id, "test-tool");
    assert_eq!(tool.timeout_seconds, 30);
}

#[test]
fn test_make_tool_overrides() {
    let mut overrides = HashMap::new();
    overrides.insert("tool_id".to_string(), serde_json::json!("custom"));
    overrides.insert("timeout_seconds".to_string(), serde_json::json!(60));
    let tool = make_tool(overrides);
    assert_eq!(tool.tool_id, "custom");
    assert_eq!(tool.timeout_seconds, 60);
}
